### ISSUE TITLE
ext_pillar: git_pillar using the new syntax not working properly (sort of?)

### ISSUE THREAD

**Issue by ekristen** [saltstack/salt#39192]
### Description of Issue/Question
When using the new git_pillar syntax and running `pillar.items` on the minion, the base environment is always returned. However, if you run `test.get_opts` or run any state that uses the pillar information from the named environment the correct pillar data is available.

### Setup
```
fileserver_backend:
  - git

gitfs_remotes:
  - git+ssh://git@github.com/myorg/my-salt-states.git

ext_pillar:
  - git: __env__ git+ssh://git@github.com/myorg/my-salt-pillars.git
```

### Steps to Reproduce Issue
Have two branches in a git repo and set it as your ext_pillar, ensure that `environment` or `pillarenv` is set to the branch name that is *not* master. Place content in both or just the named branch that is not master. I do not use the base environment so in mine there is just a README.md and no top.sls file in master.

Run `salt-call pillar.items` on your minion and you'll get back zero data, or you'll see what pillars are in your master branch. If you run `salt-call test.get_opts` under the pillar data you will see the pillar data you would expect to see from your environment.

### Behavior
After much debugging on my salt-master over the weekend I discovered that it seems that when the minion runs `pillar.items` the ext_pillar gets executed twice on the salt-master, the first time, it corrects resolves the `__env__` variable to the correct environment listed in the minions config, however after it resolves the top.sls file and creates the python dictionary, ext_pillar gets called again but this time `__env__` is resolved to master, and the contents of is what is actually returned back to the minion.

### Possible Related Bugs? 
While investigating this I found several issues already that covered portions of the behavior I was seeing. I've listed them here for completeness.

* https://github.com/saltstack/salt/issues/36223
* https://github.com/saltstack/salt/issues/21342
* https://github.com/saltstack/salt/issues/39025

### Versions Report
```
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: Ubuntu 16.04 xenial
        machine: x86_64
        release: 4.4.0-57-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

**Note:** If you want the full debug logs I can provide them.

**Comment by gtmanfred** [comment]
Thank you for reporting this.  I have been able to replicate this issue in the container `gtmanfred/issues:39192` using this set of pillars https://github.com/gtmanfred/issue-39192/

```
$ docker run --rm gtmanfred/issues:39192 salt-call pillar.items saltenv=base
local:
    ----------
    key:
        base
$ docker run --rm gtmanfred/issues:39192 salt-call pillar.items saltenv=prod
local:
    ----------
    key:
        base
$ docker run --rm gtmanfred/issues:39192 cat /etc/salt/minion.d/conf.conf
file_client: local
failhard: true
ext_pillar:
  - git:
    - __env__ git://github.com/gtmanfred/issue-39192.git
```

Thanks,
Daniel

**Comment by gtmanfred** [comment]
Actually, rereading this, if I set pillarenv or environment in the /etc/salt/minion.d/conf.conf file instead of passing saltenv on the commandline, it works for me.

[root@deb9a1c7119a /]# vi /etc/salt/minion.d/conf.conf
[root@deb9a1c7119a /]# salt-call pillar.items
local:
    ----------
    key:
        prod
[root@deb9a1c7119a /]# vim /etc/salt/minion.d/conf.conf
bash: vim: command not found
[root@deb9a1c7119a /]# vi !$
vi /etc/salt/minion.d/conf.conf
[root@deb9a1c7119a /]# salt-call pillar.items
[ERROR   ] Template does not exist:
[ERROR   ] Template does not exist:
local:
    ----------
    key:
        prod

**Comment by ekristen** [comment]
That's interesting.  @gtmanfred I have both `environment` and `pillarenv` set in my minion config and this bug happens. What else can I provide to you?

**Comment by gtmanfred** [comment]
Can you provide me a docker image that reproduces the behavior you are seeing?

**Comment by ekristen** [comment]
I'm very familiar with docker so that shouldn't be a problem. Are there instructions laying around already for creating such an image already, or something I just need to do from scratch?

**Comment by gtmanfred** [comment]
I just create mine from a scratch centos container, that should be fine.  Usually easiest to make a masterless minion with one of the following somewhere in the minion conf.

```
[root@salt ~]# cat /etc/salt/minion.d/conf.conf
file_client: local
master_type: disabled
local: true
```

**Comment by ekristen** [comment]
Well that solves that problem.  I just noticed in your config you have

```
ext_pillar:
  - git:
    - __env__ git://github.com/gtmanfred/issue-39192.git
```

In mine I had 

```
ext_pillar:
  - git: __env__ git://github.com/gtmanfred/issue-39192.git
```

And apparently this makes a difference? 

I just swapped mine to the format you had and now it works. 

So if you swap your config to the format I had you'll see the bug.

**Comment by terminalmage** [comment]
@ekristen git_pillar was rewritten for 2015.8.0, and with it the configuration schema changed. The legacy git_pilllar code is used when the repo definition on the same line as the `- git:`. The rewritten git_pillar uses shared code from gitfs, and the configuration schema is different (hence @gtmanfred's example).

We have been supporting using either the legacy or the new configuration, but it is probably time that we think about removing support for the legacy code since it has been deprecated for at least a year and a half.

The difference between the legacy and new configuration is explained in the [docs](https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.git_pillar.html).

**Comment by ekristen** [comment]
I'm a bit confused, but that's ok I guess.

What "legacy" support is being supported exactly. 

I will admit I missed the subtle difference between configuring the git backend before 2015.8.0 and after. IE same line vs new line.

Why would the pillar data be available via `test_opts` and to the states from the correct even when running but not `pillar.items` when using the single line git entry with `__env__` on salt 2016.11.2?

**Comment by terminalmage** [comment]
@ekristen Like I said, git_pillar was completely rewritten. By "legacy", I'm referring to the old git_pillar code which has been unmaintained for a couple years now. When I rewrote git_pillar, I made the decision not to completely break backwards compatibility in the ext_pillar configuration, and give people a chance to update their git_pillar configs. This also gave people a way of going back to using the legacy code as the rewritten giit_pillar code matured and bugs were squashed.

So, if you leave the git pillar configuration all on a single line, the old and unmaintained git_pillar code is used to compile the pillar data. That is why it is behaving differently depending on the two configuration schemas.
